tensorflow/core/kernels/lookup_table_op.cc

```diff
@@ -626,23 +730,48 @@ class MutableDenseHashTable final : public LookupInterface {
       int64 num_probes = 0;
       while (true) {
         if (IsEqualKey(key_buckets_matrix, bucket_index, key_matrix, i)) {
-          for (int64 j = 0; j < value_size; ++j) {
-            value_buckets_matrix(bucket_index, j) =
-                SubtleMustCopyIfIntegral(value_matrix(i, j));
+          if (use_value_ptr_) {
+            const auto* src = (char*) (value_matrix.data() + i * value_size);
+            auto* dst = (char*) key_buckets_matrix.data()
+                        + bucket_index * key_bytes_ + key_size * sizeof(K);
+            memcpy(*(void**) dst, src, value_bytes_);
+          } else {
+            for (int64 j = 0; j < value_size; ++j) {
+              value_buckets_matrix(bucket_index, j) =
+                  SubtleMustCopyIfIntegral(value_matrix(i, j));
+            }
           }
           break;
         }
         if (IsEqualKey(key_buckets_matrix, bucket_index, empty_key_tensor, 0) ||
             IsEqualKey(key_buckets_matrix, bucket_index, deleted_key_tensor,
                        0)) {
           ++num_entries_;
-          for (int64 j = 0; j < key_size; ++j) {
-            key_buckets_matrix(bucket_index, j) =
-                SubtleMustCopyIfIntegral(key_matrix(i, j));
-          }
-          for (int64 j = 0; j < value_size; ++j) {
-            value_buckets_matrix(bucket_index, j) =
-                SubtleMustCopyIfIntegral(value_matrix(i, j));
+          if (use_value_ptr_) {
+            const auto* src = (const char*) (key_matrix.data() + i * key_size);
+            auto* dst =
+                (char*) key_buckets_matrix.data() + bucket_index * key_bytes_;
+            memcpy(dst, src, key_size * sizeof(K));      // key element
+
+            src = (const char*) (value_matrix.data() + i * value_size);
+            dst += key_size * sizeof(K);
+            if (*(void**) dst == nullptr) {              // reuse value chunk
```

Comment:
So looks like we keep adding to the value_buckets_vec_ but do we ever reclaim that memory? Or does it keep on increasing? Would that cause adverse performance on workloads that have a lot of Inserts and Deletes?